File: python/practice_problems/problem_11_coverage_tracker.py
"""
=============================================================================
INTERVIEW PROBLEM 11: Sensor Coverage Tracker
Difficulty: Senior Software Engineer | Estimated time: 45 min
=============================================================================

CONTEXT
-------
You're building the health-monitoring subsystem for a platform that deploys
radio-receiver sensor stations across geographic regions. Each station
periodically sends a heartbeat. When a station falls silent, operators need
to know, and the platform's incident-detection coverage for that region may
be affected.

For this problem you are building a CoverageTracker class.
Store all state in instance variables initialized in `__init__`.
Class-level variables will bleed between tests and between CoverageTracker
instances — avoid them.
You choose the internal data structures; the public interface is what matters.

DATA MODEL
----------
Station:
  {
    "station_id": str,
    "name":       str,
    "region":     str,  # logical grouping, e.g. "downtown", "sector-7"
  }

Outage:
  {
    "station_id": str,
    "start_ts":   str,        # ISO-8601 when the outage began
    "end_ts":     str | None, # ISO-8601 when the outage ended; None = ongoing
  }

Timestamps are ISO-8601 strings without timezone offset (e.g.
"2024-01-01T10:00:00"). Use datetime.fromisoformat() for arithmetic.

# Example
# ct = CoverageTracker()
# ct.register_station("sta-001", "North Tower", region="downtown")
# ct.register_station("sta-002", "South Tower", region="downtown")
# ct.record_heartbeat("sta-001", "2024-01-01T10:00:00")
# ct.record_heartbeat("sta-002", "2024-01-01T10:00:05")
# ct.get_last_heartbeat("sta-001")              # -> "2024-01-01T10:00:00"
# ct.get_stale_stations("2024-01-01T10:05:00", stale_after_secs=120)  # -> []
# ct.record_outage_start("sta-001", "2024-01-01T10:10:00")
# ct.get_region_coverage("downtown", "2024-01-01T10:10:30", stale_after_secs=120)
# # -> {"region": "downtown", "total": 2, "healthy": 1, "stale": 1,
# #     "has_coverage": True}
=============================================================================
"""

from typing import Optional


class CoverageTracker:
    def __init__(self):
        raise NotImplementedError

    # -------------------------------------------------------------------------
    # PART 1 — Station registration and heartbeats  (~10 min)
    # -------------------------------------------------------------------------

    def register_station(self, station_id: str, name: str, region: str) -> dict:
        """
        Register a new station and return it.
        Raise ValueError if station_id already exists.
        """
        raise NotImplementedError

    def record_heartbeat(self, station_id: str, ts: str) -> None:
        """
        Record a heartbeat for the station.
        - Raise KeyError   if station_id does not exist.
        - Raise ValueError if ts is earlier than or equal to the station's most
          recent heartbeat (out-of-order and duplicate heartbeats are rejected).
        """
        raise NotImplementedError

    def get_last_heartbeat(self, station_id: str) -> Optional[str]:
        """
        Return the timestamp of the most recent heartbeat, or None if the
        station has never sent one.
        Raise KeyError if station_id does not exist.
        """
        raise NotImplementedError

    def get_stations(self, region: Optional[str] = None) -> list:
        """
        Return all stations, optionally filtered to a specific region.
        Sorted by station_id ascending.
        """
        raise NotImplementedError

    # -------------------------------------------------------------------------
    # PART 2 — Staleness detection and outage tracking  (~15 min)
    # -------------------------------------------------------------------------

    def get_stale_stations(self, as_of_ts: str, stale_after_secs: int) -> list:
        """
        Return station dicts for all stations that are stale as of as_of_ts.
        A station is stale if:
          - It has never sent a heartbeat, OR
          - Its last heartbeat was more than stale_after_secs seconds before
            as_of_ts  (i.e. as_of_ts - last_heartbeat > stale_after_secs).

        Results are sorted by station_id ascending.
        """
        raise NotImplementedError

    def record_outage_start(self, station_id: str, ts: str) -> None:
        """
        Open a new outage record for the station (end_ts = None).
        - Raise KeyError   if station_id does not exist.
        - Raise ValueError if the station already has an open outage
          (an outage with end_ts = None).
        """
        raise NotImplementedError

    def record_outage_end(self, station_id: str, ts: str) -> None:
        """
        Close the most recent open outage for the station by setting its
        end_ts = ts.
        - Raise KeyError   if station_id does not exist.
        - Raise ValueError if the station has no open outage.
        """
        raise NotImplementedError

    def get_outages(self, station_id: str) -> list:
        """
        Return all Outage dicts for the station, sorted by start_ts ascending.
        Raise KeyError if station_id does not exist.
        """
        raise NotImplementedError

    # -------------------------------------------------------------------------
    # PART 3 — Coverage analysis  (~20 min)
    # -------------------------------------------------------------------------

    def get_region_coverage(
        self, region: str, as_of_ts: str, stale_after_secs: int
    ) -> dict:
        """
        Return a coverage summary for the region:
          {
            "region":       str,
            "total":        int,   # stations in this region
            "healthy":      int,   # stations NOT stale
            "stale":        int,   # stations that ARE stale
            "has_coverage": bool,  # True if healthy >= 1
          }

        Use get_stations (Part 1) and get_stale_stations (Part 2) internally.
        """
        raise NotImplementedError

    def get_outage_summary(self, station_id: str, as_of_ts: str) -> dict:
        """
        Return an outage summary for the station:
          {
            "station_id":        str,
            "total_outages":     int,   # number of outage records
            "open_outage":       bool,  # True if there is a current open outage
            "total_outage_secs": int,   # cumulative outage duration in seconds
          }

        For an open outage (end_ts is None), count duration from start_ts up
        to as_of_ts.

        Use get_outages (Part 2) internally.
        Raise KeyError if station_id does not exist.
        """
        raise NotImplementedError

File: typescript/practice_problems/problem_11_coverage_tracker.ts
/**
 * =============================================================================
 * INTERVIEW PROBLEM 11: Sensor Coverage Tracker
 * Difficulty: Senior Software Engineer | Estimated time: 45 min
 * =============================================================================
 *
 * CONTEXT
 * -------
 * You're building the health-monitoring subsystem for a platform that deploys
 * radio-receiver sensor stations across geographic regions. Each station
 * periodically sends a heartbeat. When a station falls silent, operators need
 * to know, and the platform's incident-detection coverage for that region may
 * be affected.
 *
 * For this problem you are building a CoverageTracker class.
 * Store all state in instance properties initialized in the constructor.
 * Class-level (static) fields will bleed between tests and between
 * CoverageTracker instances — avoid them.
 * You choose the internal data structures; the public interface is what matters.
 *
 * DATA MODEL
 * ----------
 * Station:
 *   {
 *     stationId: string,
 *     name:      string,
 *     region:    string,  // logical grouping, e.g. "downtown", "sector-7"
 *   }
 *
 * Outage:
 *   {
 *     stationId: string,
 *     startTs:   string,             // ISO-8601 when the outage began
 *     endTs:     string | undefined, // ISO-8601 when the outage ended; undefined = ongoing
 *   }
 *
 * Timestamps are ISO-8601 strings without timezone offset (e.g.
 * "2024-01-01T10:00:00"). Use `new Date(ts).getTime()` for arithmetic — plain
 * string comparison also sorts them correctly since they're zero-padded.
 *
 * # Example
 * const ct = new CoverageTracker();
 * ct.registerStation("sta-001", "North Tower", "downtown");
 * ct.registerStation("sta-002", "South Tower", "downtown");
 * ct.recordHeartbeat("sta-001", "2024-01-01T10:00:00");
 * ct.recordHeartbeat("sta-002", "2024-01-01T10:00:05");
 * ct.getLastHeartbeat("sta-001");              // -> "2024-01-01T10:00:00"
 * ct.getStaleStations("2024-01-01T10:05:00", 120);  // -> []
 * ct.recordOutageStart("sta-001", "2024-01-01T10:10:00");
 * ct.getRegionCoverage("downtown", "2024-01-01T10:10:30", 120);
 * // -> { region: "downtown", total: 2, healthy: 1, stale: 1, hasCoverage: true }
 * =============================================================================
 */

export type Station = {
  stationId: string;
  name: string;
  region: string;
};

export type Outage = {
  stationId: string;
  startTs: string;
  endTs: string | undefined;
};

export type RegionCoverage = {
  region: string;
  total: number;
  healthy: number;
  stale: number;
  hasCoverage: boolean;
};

export type OutageSummary = {
  stationId: string;
  totalOutages: number;
  openOutage: boolean;
  totalOutageSecs: number;
};

export class CoverageTracker {
  constructor() {
    throw new Error("Not implemented");
  }

  // ---------------------------------------------------------------------------
  // PART 1 — Station registration and heartbeats  (~10 min)
  // ---------------------------------------------------------------------------

  /**
   * Register a new station and return it.
   * Throws an Error if stationId already exists.
   */
  registerStation(stationId: string, name: string, region: string): Station {
    throw new Error("Not implemented");
  }

  /**
   * Record a heartbeat for the station.
   * - Throws an Error if stationId does not exist.
   * - Throws an Error if ts is earlier than or equal to the station's most
   *   recent heartbeat (out-of-order and duplicate heartbeats are rejected).
   */
  recordHeartbeat(stationId: string, ts: string): void {
    throw new Error("Not implemented");
  }

  /**
   * Return the timestamp of the most recent heartbeat, or undefined if the
   * station has never sent one.
   * Throws an Error if stationId does not exist.
   */
  getLastHeartbeat(stationId: string): string | undefined {
    throw new Error("Not implemented");
  }

  /**
   * Return all stations, optionally filtered to a specific region.
   * Sorted by stationId ascending.
   */
  getStations(region?: string): Station[] {
    throw new Error("Not implemented");
  }

  // ---------------------------------------------------------------------------
  // PART 2 — Staleness detection and outage tracking  (~15 min)
  // ---------------------------------------------------------------------------

  /**
   * Return stations that are stale as of asOfTs.
   * A station is stale if:
   *   - It has never sent a heartbeat, OR
   *   - Its last heartbeat was more than staleAfterSecs seconds before
   *     asOfTs (i.e. asOfTs - lastHeartbeat > staleAfterSecs).
   *
   * Results are sorted by stationId ascending.
   */
  getStaleStations(asOfTs: string, staleAfterSecs: number): Station[] {
    throw new Error("Not implemented");
  }

  /**
   * Open a new outage record for the station (endTs = undefined).
   * - Throws an Error if stationId does not exist.
   * - Throws an Error if the station already has an open outage
   *   (an outage with endTs = undefined).
   */
  recordOutageStart(stationId: string, ts: string): void {
    throw new Error("Not implemented");
  }

  /**
   * Close the most recent open outage for the station by setting its
   * endTs = ts.
   * - Throws an Error if stationId does not exist.
   * - Throws an Error if the station has no open outage.
   */
  recordOutageEnd(stationId: string, ts: string): void {
    throw new Error("Not implemented");
  }

  /**
   * Return all Outage records for the station, sorted by startTs ascending.
   * Throws an Error if stationId does not exist.
   */
  getOutages(stationId: string): Outage[] {
    throw new Error("Not implemented");
  }

  // ---------------------------------------------------------------------------
  // PART 3 — Coverage analysis  (~20 min)
  // ---------------------------------------------------------------------------

  /**
   * Return a coverage summary for the region:
   *   { region, total, healthy, stale, hasCoverage }
   * where hasCoverage is true if healthy >= 1.
   *
   * Use getStations (Part 1) and getStaleStations (Part 2) internally.
   */
  getRegionCoverage(region: string, asOfTs: string, staleAfterSecs: number): RegionCoverage {
    throw new Error("Not implemented");
  }

  /**
   * Return an outage summary for the station:
   *   { stationId, totalOutages, openOutage, totalOutageSecs }
   *
   * For an open outage (endTs is undefined), count duration from startTs up
   * to asOfTs.
   *
   * Use getOutages (Part 2) internally.
   * Throws an Error if stationId does not exist.
   */
  getOutageSummary(stationId: string, asOfTs: string): OutageSummary {
    throw new Error("Not implemented");
  }
}
